### PR TITLE
add iter-from for bags-list

### PR DESCRIPTION
**Pull request by kianenigma**
Small backport needed in #9511 
polkadot companion: https://github.com/paritytech/polkadot/pull/5209

### CODE REVIEW

File: frame/staking/src/pallet/impls.rs [hunk 1]
```diff
@@ -1308,14 +1308,30 @@ impl<T: Config> SortedListProvider<T::AccountId> for UseNominatorsAndValidatorsM
...
 	}
+	fn iter_from(
+		start: &T::AccountId,
+	) -> Result<Box<dyn Iterator<Item = T::AccountId>>, Self::Error> {
+		if Validators::<T>::contains_key(start) {
+			let start_key = Validators::<T>::hashed_key_for(start);
+			Ok(Box::new(
+				Validators::<T>::iter_from(start_key)
+					.map(|(n, _)| n)
+					.chain(Nominators::<T>::iter().map(|(x, _)| x)),
```
Comment (ggwpez) [inline frame/staking/src/pallet/impls.rs:1326]: So you can not be a validator and nominator at the same time?

File: frame/bags-list/src/tests.rs [hunk 2]
```diff
@@ -458,6 +458,26 @@ mod sorted_list_provider {
...
+				vec![(10, vec![1, 5]), (20, vec![6]), (1000, vec![2, 3, 4])]
+			);
+
+			assert_eq!(BagsList::iter_from(&2).unwrap().collect::<Vec<_>>(), vec![3, 4, 6, 1, 5]);
+			assert_eq!(BagsList::iter_from(&3).unwrap().collect::<Vec<_>>(), vec![4, 6, 1, 5]);
+			assert_eq!(BagsList::iter_from(&4).unwrap().collect::<Vec<_>>(), vec![6, 1, 5]);
+			assert_eq!(BagsList::iter_from(&6).unwrap().collect::<Vec<_>>(), vec![1, 5]);
+			assert_eq!(BagsList::iter_from(&1).unwrap().collect::<Vec<_>>(), vec![5]);
+			assert!(BagsList::iter_from(&5).unwrap().collect::<Vec<_>>().is_empty());
+
```
Comment (ggwpez) [inline frame/bags-list/src/tests.rs:476]:
```suggestion
			assert!(BagsList::iter_from(&7).is_err());

```

File: frame/staking/src/pallet/impls.rs [hunk 3]
```diff
@@ -1308,14 +1308,30 @@ impl<T: Config> SortedListProvider<T::AccountId> for UseNominatorsAndValidatorsM
...
+			))
+		} else if Nominators::<T>::contains_key(start) {
+			let start_key = Nominators::<T>::hashed_key_for(start);
+			Ok(Box::new(Nominators::<T>::iter_from(start_key).map(|(n, _)| n)))
+		} else {
+			Err(())
+		}
+	}
```
Comment (ggwpez) [inline frame/staking/src/pallet/impls.rs:1334]: Could you maybe also add a minimal test here?

**Review by ggwpez — APPROVED** [review]

File: frame/staking/src/pallet/impls.rs (hunk 1 above)
Comment (shawntabrizi) [reply frame/staking/src/pallet/impls.rs:1326]: No, each account can have one role.

**Review by shawntabrizi — APPROVED** [review]

**Review by emostov — APPROVED** [review]
LGTM

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**